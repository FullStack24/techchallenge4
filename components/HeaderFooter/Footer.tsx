import React, { useState, useContext } from 'react';
import { View, TouchableOpacity, StyleSheet, Text, FlatList } from 'react-native';
import Icon from 'react-native-vector-icons/FontAwesome';
import { useRouter } from 'expo-router';
import useAuth from '@/hooks/useAuth';

interface FooterProps {
    userRole: string | null;
    onLogout: () => void;
    onNavigateTo: (screen: string) => void;
}

const Footer: React.FC<FooterProps> = ({ userRole, onLogout, onNavigateTo }) => {
    const [menuVisible, setMenuVisible] = useState(false);
    const { isAuthenticated } = useAuth();
    const router = useRouter();

    const options = userRole === 'teacher' ? [
        { label: 'Sala do Professor', value: '/teachers/ListTeachersScreen' },
        { label: 'Sala do Aluno', value: '/students/ListStudentsScreen' },
        { label: 'Sala do Admin', value: '/admin/AdminScreen' },
    ] : [];

    const handleHomeNavigation = () => {
        if (isAuthenticated) {
            router.push('/home/HomeScreen');
        } else {
            router.push('/auth/LoginScreen');
        }
    };

    return (
        <View style={styles.bottomNav}>
            {userRole === 'teacher' && (
                <TouchableOpacity onPress={() => setMenuVisible(!menuVisible)} style={styles.menuButton}>
                    <Icon name="book" size={24} color="#FFFFFF" />
                    <Text style={styles.navText}>Salas</Text>
                </TouchableOpacity>
            )}
            {menuVisible && (
                <View style={styles.menu}>
                    <FlatList
                        data={options}
                        renderItem={({ item }) => (
                            <TouchableOpacity
                                onPress={() => {
                                    onNavigateTo(item.value);
                                    setMenuVisible(false);
                                }}
                                style={styles.menuOption}
                            >
                                <Text style={styles.menuText}>{item.label}</Text>
                            </TouchableOpacity>
                        )}
                        keyExtractor={item => item.value}
                    />
                </View>
            )}
            <TouchableOpacity onPress={handleHomeNavigation} style={styles.navItem}>
                <Icon name="home" size={24} color="#FFFFFF" />
                <Text style={styles.navText}>Home</Text>
            </TouchableOpacity>
            <TouchableOpacity onPress={onLogout} style={styles.navItem}>
                <Icon name="sign-out" size={24} color="#FFFFFF" />
                <Text style={styles.navText}>Logout</Text>
            </TouchableOpacity>
        </View>
    );
};

const styles = StyleSheet.create({
    bottomNav: {
        flexDirection: 'row',
        backgroundColor: '#FF6B6B',
        paddingVertical: 5,
        paddingBottom: 15,
        justifyContent: 'space-around',
        alignItems: 'flex-end',
        width: '100%',
        position: 'absolute',
        bottom: 0,
        left: 0,
        right: 0,
    },
    navItem: {
        alignItems: 'center',
    },
    menuButton: {
        alignItems: 'center',
    },
    navText: {
        color: '#FFFFFF',
        fontSize: 12,
        marginTop: 2,
    },
    menu: {
        backgroundColor: '#FFFFFF',
        borderRadius: 5,
        position: 'absolute',
        bottom: 50,
        left: 10,
        right: 10,
        padding: 10,
        elevation: 5,
        zIndex: 1000,
    },
    menuOption: {
        paddingVertical: 10,
        paddingHorizontal: 15,
    },
    menuText: {
        color: '#000000',
        fontSize: 16,
        textAlign: 'left',
    },
});

export default Footer;