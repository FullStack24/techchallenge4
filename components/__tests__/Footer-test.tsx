import * as React from 'react';
import renderer from 'react-test-renderer';
import Footer from '../HeaderFooter/Footer';
import { TouchableOpacity } from 'react-native';

const onHomeMock = jest.fn();
const onLogoutMock = jest.fn();
const onNavigateToMock = jest.fn();

describe('Footer', () => {
  it('deve renderizar corretamente para um usuário com papel de "teacher"', () => {
    const tree = renderer.create(
      <Footer
        userRole="teacher"
        onHome={onHomeMock}
        onLogout={onLogoutMock}
        onNavigateTo={onNavigateToMock}
      />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('deve renderizar corretamente para um usuário sem papel de "teacher"', () => {
    const tree = renderer.create(
      <Footer
        userRole={null}
        onHome={onHomeMock}
        onLogout={onLogoutMock}
        onNavigateTo={onNavigateToMock}
      />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('deve chamar "onHome" ao clicar no botão "Home"', () => {
    const component = renderer.create(
      <Footer
        userRole="teacher"
        onHome={onHomeMock}
        onLogout={onLogoutMock}
        onNavigateTo={onNavigateToMock}
      />
    );
    const homeButton = component.root.findByType(TouchableOpacity).findByProps({ children: 'Home' });
    homeButton.props.onPress();
    expect(onHomeMock).toHaveBeenCalledTimes(1);
  });

  it('deve chamar "onLogout" ao clicar no botão "Logout"', () => {
    const component = renderer.create(
      <Footer
        userRole="teacher"
        onHome={onHomeMock}
        onLogout={onLogoutMock}
        onNavigateTo={onNavigateToMock}
      />
    );
    const logoutButton = component.root.findByType(TouchableOpacity).findByProps({ children: 'Logout' });
    logoutButton.props.onPress();
    expect(onLogoutMock).toHaveBeenCalledTimes(1);
  });

  it('deve alternar a visibilidade do menu de navegação quando clicar no botão "Salas"', () => {
    const component = renderer.create(
      <Footer
        userRole="teacher"
        onHome={onHomeMock}
        onLogout={onLogoutMock}
        onNavigateTo={onNavigateToMock}
      />
    );

    const menuButton = component.root.findByType(TouchableOpacity).findByProps({ children: 'Salas' });
    menuButton.props.onPress();
    expect(component.root.findByType('FlatList')).toBeTruthy();

    menuButton.props.onPress();
    expect(() => component.root.findByType('FlatList')).toThrow();
  });
});
