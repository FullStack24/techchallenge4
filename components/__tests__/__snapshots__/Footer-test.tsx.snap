// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer deve renderizar corretamente para um usuário com papel de "teacher" 1`] = `
<View
  style={
    {
      "alignItems": "flex-end",
      "backgroundColor": "#FF6B6B",
      "bottom": 0,
      "flexDirection": "row",
      "justifyContent": "space-around",
      "left": 0,
      "paddingBottom": 15,
      "paddingVertical": 5,
      "position": "absolute",
      "right": 0,
      "width": "100%",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "opacity": 1,
      }
    }
  >
    <Text />
    <Text
      style={
        {
          "color": "#FFFFFF",
          "fontSize": 12,
          "marginTop": 2,
        }
      }
    >
      Salas
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "opacity": 1,
      }
    }
  >
    <Text />
    <Text
      style={
        {
          "color": "#FFFFFF",
          "fontSize": 12,
          "marginTop": 2,
        }
      }
    >
      Home
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "opacity": 1,
      }
    }
  >
    <Text />
    <Text
      style={
        {
          "color": "#FFFFFF",
          "fontSize": 12,
          "marginTop": 2,
        }
      }
    >
      Logout
    </Text>
  </View>
</View>
`;

exports[`Footer deve renderizar corretamente para um usuário sem papel de "teacher" 1`] = `
<View
  style={
    {
      "alignItems": "flex-end",
      "backgroundColor": "#FF6B6B",
      "bottom": 0,
      "flexDirection": "row",
      "justifyContent": "space-around",
      "left": 0,
      "paddingBottom": 15,
      "paddingVertical": 5,
      "position": "absolute",
      "right": 0,
      "width": "100%",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "opacity": 1,
      }
    }
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": "#FFFFFF",
            "fontSize": 24,
          },
          undefined,
          {
            "fontFamily": "FontAwesome",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
    >
      
    </Text>
    <Text
      style={
        {
          "color": "#FFFFFF",
          "fontSize": 12,
          "marginTop": 2,
        }
      }
    >
      Home
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "opacity": 1,
      }
    }
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": "#FFFFFF",
            "fontSize": 24,
          },
          undefined,
          {
            "fontFamily": "FontAwesome",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
    >
      
    </Text>
    <Text
      style={
        {
          "color": "#FFFFFF",
          "fontSize": 12,
          "marginTop": 2,
        }
      }
    >
      Logout
    </Text>
  </View>
</View>
`;
